cps/elf.c

```diff
@@ -0,0 +1,267 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <elf.h>
+
+#include "gatekeeper_cps.h"
+#include "elf.h"
+
+#if __WORDSIZE == 32
+#define ElfPERBIT(x) Elf32_##x
+#else
+#define ElfPERBIT(x) Elf64_##x
+#endif
+
+struct elf_file {
+	/* File contents. */
+	const void    *data;
+
+	/* Length of file contents. */
+	unsigned long len;
+
+	/* Whether the endianness needs to be converted. */
+	int           conv;
+};
+
+/*
+ * Get CPU endianness.
+ *
+ * 1 = ELFDATA2LSB = little
+ * 2 = ELFDATA2MSB = big
+ */
+static int __attribute__((pure))
+native_endianness(void)
+{
+	/*
+	 * Encoding the endianness enums in a string
+	 * and then reading that string as a 16-bit integer.
+	 */
+	return (char)*((const uint16_t *)("\1\2"));
+}
+
+static inline void
+__swap_bytes(const void *src, void *dest, unsigned int size)
+{
+	unsigned int i;
+	for (i = 0; i < size; i++) {
+		((unsigned char *)dest)[i] =
+			((const unsigned char *)src)[size - i - 1];
+	}
+}
+
+/* Change endianness of @x if @conv is true storing it in @buf_ptr. */
+#define END(x, conv, buf_ptr)						\
+({									\
+	if (conv) __swap_bytes(&(x), (buf_ptr), sizeof(*buf_ptr));	\
+	else *(buf_ptr) = (x);						\
+})
+
+/* Get the section specified in @secname from the ELF file. */
+static const void *
+get_section(const struct elf_file *mod, const char *secname,
+	unsigned long *secsize)
+{
+	unsigned long len = mod->len;
+	int conv = mod->conv;
+	unsigned int i;
+	const char *secstrings;
+
+	const ElfPERBIT(Ehdr) *hdr = mod->data;
+	const ElfPERBIT(Shdr) *sechdrs;
+	ElfPERBIT(Off) e_shoff_ntv;
+	ElfPERBIT(Half) e_shnum_ntv;
+	ElfPERBIT(Half) e_shstrndx_ntv;
+	ElfPERBIT(Off) sh_offset_ntv;
+
+	*secsize = 0;
+
+	/* Check whether there's enough room for ELF header. */
+	if (len < sizeof(*hdr))
+		return NULL;
+
+	/* Read in necessary variables from ELF header and fix endianness. */
+	END(hdr->e_shoff, conv, &e_shoff_ntv);
+	END(hdr->e_shnum, conv, &e_shnum_ntv);
+	END(hdr->e_shstrndx, conv, &e_shstrndx_ntv);
+
+	/* Check whether there's enough room for the section headers. */
+	if (len < e_shoff_ntv + e_shnum_ntv * sizeof(sechdrs[0]))
+		return NULL;
+
+	sechdrs = (const ElfPERBIT(Shdr) *)
+		((const uint8_t *)hdr + e_shoff_ntv);
+
+	/* Check whether there's enough room for section with section names. */
+	END(sechdrs[e_shstrndx_ntv].sh_offset, conv, &sh_offset_ntv);
+	if (len < sh_offset_ntv)
+		return NULL;
+
+	/* Find symbol table. */
+	secstrings = (const char *)((const uint8_t *)hdr + sh_offset_ntv);
+
+	for (i = 1; i < e_shnum_ntv; i++) {
+		ElfPERBIT(Word) sh_name_ntv;
+		const char *found_secname;
+
+		END(sechdrs[i].sh_name, conv, &sh_name_ntv);
+		found_secname = secstrings + sh_name_ntv;
+
+		if (strcmp(secname, found_secname) == 0) {
+			ElfPERBIT(Word) sh_size_ntv;
```

Comment:
Once `sh_size_ntv` is declared earlier due to another comment, we don't need to declare it again here.